Index: b/Makefile.am
===================================================================
--- a/Makefile.am	2008-07-12 08:57:59.299247757 -0400
+++ b/Makefile.am	2008-07-12 08:58:51.541600527 -0400
@@ -3,7 +3,7 @@
 AUTOMAKE_OPTIONS = foreign subdir-objects
 
 AM_CFLAGS        = ${regular_CFLAGS} -I${top_builddir}/include -I${top_srcdir}/include ${kinclude_CFLAGS}
-SUBDIRS         := extensions
+SUBDIRS         := extensions howtos
 if ENABLE_LIBIPQ
 SUBDIRS         += libipq
 endif
@@ -102,3 +102,6 @@
 	pushd /tmp/${PACKAGE_TARNAME}-${PACKAGE_VERSION} && ./autogen.sh && popd;
 	tar -C /tmp -cjf ${PACKAGE_TARNAME}-${PACKAGE_VERSION}.tar.bz2 --owner=root --group=root ${PACKAGE_TARNAME}-${PACKAGE_VERSION}/;
 	rm -Rf /tmp/${PACKAGE_TARNAME}-${PACKAGE_VERSION};
+
+check: recursive-check
+	echo "makefile jedi handwaving"
